### ISSUE TITLE
MSL: gl_TessCoord mis-declared as a float2 instead of correct float3

### ISSUE THREAD

**Issue by billhollings** [KhronosGroup/SPIRV-Cross#1762]
SPIR-V conversion to MSL fails compilation due to `gl_TessCoord` mis-declared as a `float2` instead of correct `float3`:

These CTS test fail as a result:
```
dEQP-VK.spirv_assembly.instruction.graphics.cross_stage.basic_type.flat
dEQP-VK.spirv_assembly.instruction.graphics.cross_stage.basic_type.no_perspective
dEQP-VK.spirv_assembly.instruction.graphics.cross_stage.basic_type.relaxedprecision
dEQP-VK.spirv_assembly.instruction.graphics.cross_stage.interface_blocks.flat
dEQP-VK.spirv_assembly.instruction.graphics.cross_stage.interface_blocks.no_perspective
dEQP-VK.spirv_assembly.instruction.graphics.cross_stage.interface_blocks.relaxedprecision
```
SPIR-V file: 
[gl_TessCoord_as_float2_bad.spv.zip](https://github.com/KhronosGroup/SPIRV-Cross/files/7304932/gl_TessCoord_as_float2_bad.spv.zip)

**Comment by etra0** [comment]
Hmm, from what I can see, the only criteria to choose between `float2` and `float3` is the execution mode:
https://github.com/KhronosGroup/SPIRV-Cross/blob/e9cc6403341baf0edd430a4027b074d0a06b782f/spirv_msl.cpp#L14583-L14584

Looking at the uploaded SPIR-V and a test that expects a `float2` instead of a `float3` (for example: https://github.com/KhronosGroup/SPIRV-Cross/blob/master/shaders-msl/asm/tese/unnamed-builtin-array.asm.tese) I can't see a specific factor besides the index access and the execution mode being `SpacingFractionalEven` vs the uploaded file which is `SpacingEqual`. 

Adding the `|| execution.flags.get(ExecutionModeSpacingEqual)` to the condition makes all test pass with the uploaded file included but I **seriously** doubt that's the correct way of doing it, and I don't have much experience with MSL either.

Do you have any idea what other criteria could be used to choose the correct type?

**Comment by cfnptr** [comment]
Looks like the problem is that the SPIRV-MSL translator sets all three `gl_TessCoord` values (x, y, z) no matter what type of patch (Triangles or Quads) is used in the shader.

As we can see from the decompiled [gl_TessCoord_as_float2_bad.spv.zip](https://github.com/KhronosGroup/SPIRV-Cross/files/7304932/gl_TessCoord_as_float2_bad.spv.zip) shader, variable `float _21` isn't used in calculations.

```
#version 450
layout(quads, ccw, equal_spacing) in;

layout(location = 0) out vec4 _27;
layout(location = 0) in vec4 _37[];
layout(location = 1) out float _71;
layout(location = 1) in float _79[];
layout(location = 2) out vec2 _108;
layout(location = 2) in vec2 _116[];
layout(location = 3) out vec3 _145;
layout(location = 3) in vec3 _153[];
layout(location = 4) out vec4 _180;
layout(location = 4) in vec4 _186[];

void main()
{
    float _8 = gl_TessCoord.x;
    float _17 = gl_TessCoord.y;
    float _21 = gl_TessCoord.z;
    _27 = (((_37[0] * ((1.0 - _8) * (1.0 - _17))) + (_37[1] * ((1.0 - _8) * _17))) + (_37[2] * (_8 * (1.0 - _17)))) + (_37[3] * (_8 * _17));
    _71 = (((((1.0 - _8) * (1.0 - _17)) * _79[0]) + (((1.0 - _8) * _17) * _79[1])) + ((_8 * (1.0 - _17)) * _79[2])) + ((_8 * _17) * _79[3]);
    _108 = (((_116[0] * ((1.0 - _8) * (1.0 - _17))) + (_116[1] * ((1.0 - _8) * _17))) + (_116[2] * (_8 * (1.0 - _17)))) + (_116[3] * (_8 * _17));
    _145 = (((_153[0] * ((1.0 - _8) * (1.0 - _17))) + (_153[1] * ((1.0 - _8) * _17))) + (_153[2] * (_8 * (1.0 - _17)))) + (_153[3] * (_8 * _17));
    _180 = (((_186[0] * ((1.0 - _8) * (1.0 - _17))) + (_186[1] * ((1.0 - _8) * _17))) + (_186[2] * (_8 * (1.0 - _17)))) + (_186[3] * (_8 * _17));
    gl_Position = (((gl_in[0].gl_Position * ((1.0 - _8) * (1.0 - _17))) + (gl_in[1].gl_Position * ((1.0 - _8) * _17))) + (gl_in[2].gl_Position * (_8 * (1.0 - _17)))) + (gl_in[3].gl_Position * (_8 * _17));
}
```

And this condition ```execution.flags.get(ExecutionModeTriangles) ? "float3" : "float2"``` apparently exists because as declared [here](https://developer.apple.com/library/archive/documentation/Miscellaneous/Conceptual/MetalProgrammingGuide/Tessellation/Tessellation.html), "...For quad patches, must be float2. For triangle patches, must be float3".

I think that it's more correct to fix the translator and not set the z value if it's not used.

**Comment by etra0** [comment]
Hmm, although is not used in the calculations, in the disassembly it's declared as a `float3` and does an `OpAccessChain` to the third element:
```asm
    %v3float = OpTypeVector %float 3
%_ptr_Input_v3float = OpTypePointer Input %v3float
%gl_TessCoord = OpVariable %_ptr_Input_v3float Input
;              (...)
         %15 = OpAccessChain %_ptr_Input_float %gl_TessCoord %uint_0
         %16 = OpLoad %float %15
               OpStore %8 %16
         %19 = OpAccessChain %_ptr_Input_float %gl_TessCoord %uint_1
         %20 = OpLoad %float %19
               OpStore %17 %20
         %23 = OpAccessChain %_ptr_Input_float %gl_TessCoord %uint_2
         %24 = OpLoad %float %23
               OpStore %21 %24
```
so I'm not entirely sure to just dismiss the third access.

**Comment by cfnptr** [comment]
Hmm, interesting. This is probably because `gl_TessCoord` in GLSL has only `float3` variant, but I'm too not sure.
@billhollings, should the translated MSL shaders contain exactly the same commands as in SPIRV?

**Comment by billhollings** [comment]
Hmmm...I wonder if we can fix this in MSL by creating something like `gl_TessCoordIn` as a synthetic Input variable, either as a `float2` or `float3` as appropriate, and then assigning it to `gl_TessCoord`, at the beginning of the shader `main()` function, populating `gl_TessCoord.z` with `0.0` if `gl_TessCoordIn` has only two components.

The rest of the shader would then use `gl_TessCoord` as a 3-component variable, as expected.

**Comment by billhollings** [comment]
Fixed in #1824.
